iree/turbine/kernel/wave/compile.py

```diff
@@ -32,15 +34,25 @@ def invoke(self, *args, **kwargs):
         Returns the assembly code of the compiled kernel.
         """
 
+        # Segregate args into kernel tensor and scalars.
+        tensor_args, scalar_args = [], []
+        kernel_inputs, kernel_outputs = [], []
+
         # Partition arguments into kernel inputs and outputs.
-        kernel_inputs = []
-        kernel_outputs = []
-        for arg, usage in zip(args, self.options.kernel_usages):
-            if usage == kernel_codegen.KernelBufferUsage.INPUT:
-                kernel_inputs.append(arg)
-
-            if usage == kernel_codegen.KernelBufferUsage.OUTPUT:
-                kernel_outputs.append(arg)
+        usage_idx = 0
+        for arg in args:
+            if isinstance(arg, torch.Tensor):
```

Comment:
NIT 1:  tensor_args is no longer required 
NIT 2: I think IREE/MLIR style convention is try to do early exit/minimize nested, i.e, we can do something like:
```py
usage_idx = 0
        for arg in args:
            if not isinstance(arg, torch.Tensor):
               scalar_args.append(arg)
               continue
            usage = self.options.kernel_usages[usage_idx]
            usage_idx += 1
            if usage == kernel_codegen.KernelBufferUsage.INPUT:
              kernel_inputs.append(arg)
            if usage == kernel_codegen.KernelBufferUsage.OUTPUT:
              kernel_outputs.append(arg)
```

Reply:
would the scalar inputs not have `self.options.kernel_usages`?

Reply:
I see because it's `BindingType.SCALAR_VALUE` it doesn't work out of the box

Reply:
perhaps could be good to make a function called `kernel_signature_bindings` that takes generate binding for all inputs/outputs that is part of the kernel signature. That way modification to this part will be very minimal

Reply:
I agree with the early stopping but we may not want to add another function call as it is not used in many places. Also, it will be same as adding one more loop.

Reply:
Moved to a TODO, where we need to add `kernel_signature_binding` but also propagate kernelbufferusage/scalartypeusage to the bindingdesc